Issue 381: Add a command for auditing the state of the database

Hzaakk opened the pull request:
## Changes
- Added a new command via the file coldfront/core/utils/management/commands/audit_data.py
- Usage: `python manage.py audit_data --all [--lrc-all]` or `python manage.py --audit-a [--audit-b ...]`
- Arguments:
<img width="542" alt="image" src="https://user-images.githubusercontent.com/43729496/179861754-e9cdd351-279b-41c1-86f6-54c8f820605a.png">

## How I tested changes
- I ran the command, observing output from every audit (every audit, but not every case within every audit, produced output).
- I logged into my local portal and confirmed the output from the audits.
- I didn't write manual test cases to test every case.

## Notes
- I added the initial code for LRC auditing, but haven't tested the code and it probably doesn't all work.

## Usage Screenshots
<img width="830" alt="image" src="https://user-images.githubusercontent.com/43729496/179861241-dbad08de-da2a-4f3a-8e21-941704c3faf7.png">
<img width="1196" alt="image" src="https://user-images.githubusercontent.com/43729496/179861260-5793e7ce-e8dc-465a-aa75-834296559a2f.png">
<img width="788" alt="image" src="https://user-images.githubusercontent.com/43729496/179861328-07b65ef8-6846-4b01-bb37-32636b3e4b7b.png">

jofeinstein commented on coldfront/core/utils/management/commands/audit_data.py:
```diff
@@ -0,0 +1,229 @@
+        '''
+        Assert that all LRC Recharge Allocations have a "Billing Activity"-typed AllocationAttribute.
+        '''
+        allocations = Allocation.objects.filter(project__cluster__name='Lawrencium')
```
project__cluster__name is not a valid lookup. I think we need to filter on the Resource having a 'Compute' name since there's a lot of clusters on the LRC side

Hzaakk replied:
Oh thanks. I stopped working on the LRC stuff because I don't know if I'm only supposed to work on BRC and also I'm not familiar with LRC resources and allocations. I'll change the filter and add a `TODO` for every LRC command in a while.

jofeinstein commented on coldfront/core/utils/management/commands/audit_data.py:
```diff
@@ -0,0 +1,229 @@
+        ICA_PREFIX = 'ic_'
+        RECHARGE_CONDO_PREFIX = ('co_', 'ac_')
+
+        for allocation in allocations:
```
Can we order the results for this and the commands below so the output is more organized?

Hzaakk replied:
Yeah, would ordering initially by (project__name, projet__status__name, start_date, end_date) be fine, or is that too many columns to order by? Is there a way to sort by project name but in a way that treats FCAs and PCAs the same, and Condo and Recharge the same? Maybe grouping somehow?

jofeinstein replied:
I think initially ordering with what you mentioned should be fine

jofeinstein reviewed (commented):
This works for me. I added a couple of comments.

Hzaakk commented on coldfront/core/utils/management/commands/audit_data.py:
```diff
@@ -0,0 +1,229 @@
+        
+        for project in inactive_projects:
+            for allocation in project.allocation_set.select_related('status').prefetch_related('allocationattribute_set'):
+                if allocation.status.name != 'Expired':
```
It looks like all the output of this command points to inactive projects with active secure directories. If that's expected behavior then we can filter out secure directories.

jofeinstein replied:
On second thought, its probably good to include secure directories here. Could we print a different error message if the project has an active secure directory but is inactive?

Hzaakk commented on coldfront/core/utils/management/commands/audit_data.py:
```diff
@@ -0,0 +1,229 @@
+        users = UserProfile.objects.filter(cluster_uid__isnull=False) \
+            .values('id', 'user__first_name', 'user__last_name', 'user__email', 'access_agreement_signed_date')
+        for user in users:
+            user_project_count = ProjectUser.objects.filter(user_id=user['id']).count()
```
We can filter to only active users, yeah. Maybe we should also filter to users that have signed the access agreement as they make up all the output of the command (looks like all inactive users haven't signed, but some active users haven't also signed either). What do you think?

jofeinstein replied:
I'm not sure about the access agreement. Lets leave that up to Matthew

Hzaakk replied:
I just found out that since denied and removed projectusers still have an entry in the projectuser table, the exists() method returns true for them, so they were excluded by default.

matthew-li replied:
We can ignore whether they've signed the access agreement.

When we initially loaded in users, we made everyone "Active", but left the agreement unsigned.

Anyone coming to the portal needs to sign it as the first step. But there may still be some active cluster users who haven't signed the agreement because they haven't needed to come to the portal yet.

Hzaakk replied:
So does that mean no changes needed then?

jofeinstein commented on coldfront/core/utils/management/commands/audit_data.py:
```diff
@@ -0,0 +1,229 @@
+ Users with a cluster UID should be associated with at least one Project (maybe).
+LRC only:
+ Each UserProfile that has access to Lawrencium should have a billing_activity.
+ Each Recharge Allocation that has access to Lawrencium should have a "Billing Activity"-typed AllocationAttribute.
```
Can you limit line length to 80 characters? There's usually an option in your IDE to show you when you go over.

jofeinstein commented on coldfront/core/utils/management/commands/audit_data.py:
```diff
@@ -0,0 +1,229 @@
+        '''
+        projects = Project.objects.all()
+        for project in projects:
+            if project.pis().count() == 0:
```
Whenever you do queryset.count() == 0, you can also use queryset.exists(). It'll return true if the count > 0 and false otherwise

Hzaakk commented:
@jofeinstein I think I've addressed all your comments. 

Changes:
- All output is now sorted
- Error messages are better:
    - More info included
    - Statuses for everything
    - For active secure directories on inactive projects, I now use yellow warnings with a `CURRENTLY EXPECTED BEHAVIOR: ` prefix.
 
<img width="1052" alt="image" src="https://user-images.githubusercontent.com/43729496/180591253-89972df6-9274-4235-b926-add9bc596e2a.png">
<img width="1115" alt="image" src="https://user-images.githubusercontent.com/43729496/180591257-3f302f7d-adbf-4b59-ada8-c98d4fa22f51.png">

matthew-li commented on coldfront/core/utils/management/commands/audit_data.py:
```diff
@@ -0,0 +1,313 @@
+
+        FCA_PCA_ALLOCATION_PERIOD = AllocationPeriod.objects.get(
+                                        name__startswith='Allowance Year',
+                                        start_date__lt=datetime.date.today(),
```
You can use `coldfront.core.project.utils_.renewal_utils.get_current_allowance_year_period` here.

matthew-li commented on coldfront/core/utils/management/commands/audit_data.py:
```diff
@@ -0,0 +1,313 @@
+                                        start_date__lt=datetime.date.today(),
+                                        end_date__gt=datetime.date.today())
+        ICA_ALLOCATION_PERIODS = AllocationPeriod.objects.filter(
+                                    Q(end_date__gt=datetime.date.today())
```
You should use `oldfront.core.utils.common.display_time_zone_current_date` to get a date that takes the time zone into account.

matthew-li commented on coldfront/core/utils/management/commands/audit_data.py:
```diff
@@ -0,0 +1,313 @@
+            resource = allocation['resources__name']
+
+            if end_date is not None and end_date < start_date:
+                self.stdout.write(self.style.ERROR('Allocation {} '
```
We've tended to use Python 3 f-strings, which are shorter, and (IMO) easier to read.

matthew-li commented on coldfront/core/utils/management/commands/audit_data.py:
```diff
@@ -0,0 +1,313 @@
+                    'that is different than its allocation period\'s ({}).' \
+                        .format(resource, id, project,
+                            start_date, FCA_PCA_ALLOCATION_PERIOD.start_date)))
+                if project_status == 'Active' \
```
We could also assert that the start date is no earlier than the start date of the allocation period.

matthew-li commented on coldfront/core/utils/management/commands/audit_data.py:
```diff
@@ -0,0 +1,313 @@
+                if end_date is not None:
+                    self.stdout.write(self.style.ERROR('{} {} '
+                    'for {} Recharge or Condo allocation project {} has '
+                    'an end date. (it shouldn\'t' \
```
There's a missing closing parenthesis here.

matthew-li commented on coldfront/core/utils/management/commands/audit_data.py:
```diff
@@ -0,0 +1,313 @@
+                              .select_related('status') \
+                              .prefetch_related('allocationattribute_set',
+                                                'resources'):
+                resource = allocation.resources.first().name
```
This produced an error on my box:

```
  File "/vagrant/coldfront_app/coldfront/coldfront/core/utils/management/commands/audit_data.py", line 199, in handle_project_inactive
    resource = allocation.resources.first().name
AttributeError: 'NoneType' object has no attribute 'name'
```

Some exception handling is in order.

matthew-li commented on coldfront/core/utils/management/commands/audit_data.py:
```diff
@@ -0,0 +1,313 @@
+                try:
+                    allocation_attribute = allocation.allocationattribute_set \
+                        .get(allocation_attribute_type__name='Service Units')
+                    if allocation_attribute.value == 0:
```
The value is a string, so I don't think this case would ever fire.

matthew-li replied:
Also, should this be `!=`?

Hzaakk replied:
Yeah, it should be != oops. I found 2 different 0 values for the attribute, "0" and "0.00". Are there any other ones I should be aware of (like None or "0.0")

matthew-li replied:
We typically try to coerce it into a `Decimal`--if that fails, then there's a corrupted value.

matthew-li commented on bootstrap/development/load_database_backup.sh:
```diff
@@ -4,26 +4,36 @@
 # loads in the data from the given dump file generated by pg_dump. It
```
Merging `develop` into this branch should remove this from the diff?

matthew-li reviewed (changes requested):
Thanks! I did a quick pass over the BRC code, which largely looks reasonable. But I ran into an error I mentioned in the comments that prevented me from running the command to completion.

matthew-li commented on coldfront/core/utils/management/commands/audit_data.py:
```diff
@@ -0,0 +1,323 @@
+                                       | Q(name__startswith='Summer Sessions') \
+                                      )).all()
+
+        # TODO: Change these if lrc_enabled
```
After this code was written, as part of readying the codebase for both deployments, I added `coldfront.core.resource.utils_.allowance_utils` to handle the differing allowance types across the two.

There's a function for getting the correct allowance for a `Project` object. Then, based on whether it's yearly or instructional, you can perform the right checks. Project prefixes shouldn't be hard-coded (except maybe in tests) moving forward.

Hzaakk replied:
I used `ComputingAllowance(allocation).is_recharge() or ComputingAllowance(allocation).has_infinite_service_units()` to account for condo and recharge, because there's no check for condo resources.

Hzaakk replied:
Actually I went ahead and added is_condo myself

matthew-li replied:
Looks good, thanks.

matthew-li commented on coldfront/core/utils/management/commands/audit_data.py:
```diff
@@ -0,0 +1,323 @@
+'''
```
FYI: Prior to launching MyLRC, I wrote `core.utils.management.commands.validate_lrc_initial_state`. It doesn't really cover the LRC checks you'll need to do here, but you can use it as a reference.

matthew-li commented on coldfront/core/utils/management/commands/audit_data.py:
```diff
@@ -0,0 +1,323 @@
+        users = UserProfile.objects.filter(cluster_uid__isnull=False) \
+            .select_related('user') \
+            .order_by('user__is_active', 'user__username') \
+            .values('id', 'user__first_name', 'user__last_name',
```
Is there a reason to fetch and use the values in this way?

Instead of `user["user__last_name"]`, you can do `user.user.last_name`. (The first `user` would be more accurately named `user_profile`.)

If the concern is database lookups, you can fetch everything you need all upfront using `prefetch_related` and `select_related`.

Hzaakk replied:
I heard that it's more efficient but if that doesn't matter I can swap it to the normal usage

matthew-li replied:
Please do, mainly for consistency with the rest of the codebase.

I've never used [`only`](https://docs.djangoproject.com/en/4.1/ref/models/querysets/#only), but it looks like it accomplishes the same thing (it only retrieves the fields you want), while still returning ORM objects.

```
>>> User.objects.first().__dict__
{'_state': <django.db.models.base.ModelState object at 0x7f9580a07128>, 'id': 1, 'password': '', 'last_login': None, 'is_superuser': False, 'username': 'username', 'first_name': 'First', 'last_name': 'Last', 'email': 'email@berkeley.edu', 'is_staff': False, 'is_active': True, 'date_joined': datetime.datetime(1970, 1, 1, 0, 0, 0, 0, tzinfo=<UTC>)}
>>> User.objects.only("first_name").first().__dict__
{'_state': <django.db.models.base.ModelState object at 0x7f9580a07898>, 'id': 1, 'first_name': 'First'}
```

matthew-li commented on coldfront/core/utils/management/commands/audit_data.py:
```diff
@@ -0,0 +1,323 @@
+                f'{user["user__last_name"]}, {user["user__email"]}) '
+                f'has a cluster UID but is not associated with any projects.'))
+
+    def handle_lrc_user_billing(self):
```
We'll also want to check validity--see `coldfront.core.billing.utils.validation.is_billing_id_valid`.

This won't be testable on a dev. VM because the billing database is restricted.

(I'm also happy to write the LRC checks, since you don't have an LRC VM--let me know.)

Hzaakk replied:
I think it'd be great if you wrote the LRC checks as I'm bound to deal with a bunch of errors trying it without a VM

matthew-li commented on coldfront/core/utils/management/commands/audit_data.py:
```diff
@@ -0,0 +1,323 @@
+            if project.startswith(FCA_PCA_PREFIXES):
+                if project_status == 'Inactive' \
+                        and start_date != FCA_PCA_ALLOCATION_PERIOD.start_date:
+                    self.stdout.write(self.style.ERROR(f'{resource} {id} '
```
Neither deployment conforms to this check at the moment, for different reasons.

On LRC, we ran the allocation renewals a bit late this year to coincide with launch. Let's apply the `<` check from L156 to `Inactive` projects as well.

On BRC, I may just end up rewriting start/end dates for inactive projects to be the correct ones.

Hzaakk replied:
I swapped the != for < if that's what you mean

matthew-li replied:
I meant that it's acceptable for both Active and Inactive projects to have start dates greater than or equal to the start date of the corresponding allocation period.

matthew-li reviewed (approved):
I'll add LRC audits in the future.
